### PR TITLE
fix a bug for futurebuilder, example:

### PR DESCRIPTION
**Pull request by xiongzhaowei**
```dart
Builder(
  builder: (context) {
    final identifier = Object();
    return FutureBuilder(
      future: Future.value(identifier),
      builder: (context, snapshot) {
        assert(identifier == snapshot.data || snapshot.data == null);
        return Container();
      },
    );
  },
)
```
The assertion will fail when the FutureBuilder is rebuilt, but it will succeed after this fix.

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by xiongzhaowei** [comment]
This is a bug I found during development, and I don't know how to submit a pull request. Please help me if you can.

**Comment by goderbauer** [comment]
Hi @xiongzhaowei It looks like this PR is missing a test. Can you please add one. Also, it looks like many checks are failing. Can you please take a look at those and address the issues? Thank you!

File: packages/flutter/lib/src/widgets/async.dart [hunk 1]
```diff
@@ -596,6 +596,8 @@ class _FutureBuilderState<T> extends State<FutureBuilder<T>> {
   /// or after widget reconfiguration to a new Future.
   Object? _activeCallbackIdentity;
   late AsyncSnapshot<T> _snapshot;
+  Future<T>? _future;
+  Widget? _lastWidget;
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/async.dart:600]: You should not caches widget in the state, that is what didUpdateWidget is for.

Can you apply the fix to didUpdateWidget instead?

**Comment by goderbauer** [comment]
(triage) @xiongzhaowei Do you still have plans to come back to this PR and address the feedback given above? Also, it seems like a few checks are failing. Can you take a look at those as well? Thank you.

**Comment by xiongzhaowei** [comment]
This is my first PR and I don't know how to get it to the required level to be accepted. I noticed that there were some linux-related checks that didn't work, but I didn't modify those.

I found a new way to fix this without caching the widget.

**Review by chunhtai — COMMENTED** [review]
I don't think this is really a bug, the doc specifically said not to create future during build
https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html
```
The [future](https://api.flutter.dev/flutter/widgets/FutureBuilder/future.html) must have been obtained earlier, e.g. during [State.initState](https://api.flutter.dev/flutter/widgets/State/initState.html), [State.didUpdateWidget](https://api.flutter.dev/flutter/widgets/State/didUpdateWidget.html), or [State.didChangeDependencies](https://api.flutter.dev/flutter/widgets/State/didChangeDependencies.html). It must not be created during the [State.build](https://api.flutter.dev/flutter/widgets/State/build.html) or [StatelessWidget.build](https://api.flutter.dev/flutter/widgets/StatelessWidget/build.html) method call when constructing the [FutureBuilder](https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html). If the [future](https://api.flutter.dev/flutter/widgets/FutureBuilder/future.html) is created at the same time as the [FutureBuilder](https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html), then every time the [FutureBuilder](https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html)'s parent is rebuilt, the asynchronous task will be restarted.
```
and also
```
A side-effect of this is that providing a new but already-completed future to a [FutureBuilder](https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html) will result in a single frame in the [ConnectionState.waiting](https://api.flutter.dev/flutter/widgets/ConnectionState.html) state. This is because there is no way to synchronously determine that a [Future](https://api.flutter.dev/flutter/dart-async/Future-class.html) has already completed.
```

if you really want to do that, then you should at least check for connection state
```dart
    testWidgets('The variable captured by FutureBuilder should be equal to the return value of the future.', (WidgetTester tester) async {
      final StreamController<Object?> streamController = StreamController<Object?>();
      await tester.pumpWidget(StreamBuilder<Object?>(
        stream: streamController.stream,
        builder: (BuildContext context, AsyncSnapshot<Object?> snapshot) {
          final Object identifier = Object();
          return FutureBuilder<Object>(
            initialData: identifier,
            future: Future<Object>.value(identifier),
            builder: (BuildContext context, AsyncSnapshot<Object> snapshot) {
              if (snapshot.connectionState == ConnectionState.active) {
                expect(identifier, snapshot.data,
                    reason: '“identifier” should be equal to “snapshot.data”.');
              }
              return Container();
            },
          );
        },
      ));
      await tester.pump();
      for (int i = 0; i < 10; i++) {
        streamController.add(null);
        await tester.pump();
      }
    });
```

**Pull request closed without merging**